src/module.c

```diff
@@ -2023,6 +2023,139 @@ int RM_PublishMessage(RedisModuleCtx *ctx, RedisModuleString *channel, RedisModu
     return receivers;
 }
 
+/* Function pointer type for subscription from modules. */
+typedef int (*RedisModuleSubscribeFunc) (RedisModuleCtx *ctx, RedisModuleString *channel,  RedisModuleString *message);
+
+/* Channel subscriber information.
+ * See for more information. */
+typedef struct RedisModuleChannelSubscriber {
+    /* The module subscribed to the event */
+    RedisModule *module;
+    /* Notification callback in the module*/
+    RedisModuleSubscribeFunc subscribe_callback;
+} RedisModuleChannelSubscriber;
+
+/* The channel to module subscribers dict. */
+static dict *channelModuleSubscribers;
+
+/* The module to channel subscribers dict. */
+static dict *moduleChannelSubscribers;
+
+int removeModuleFromChannelSubscription(const char *module, robj *channel) {
+    list *subscribers = NULL;
+    listIter li;
+    int retval;
+    dictEntry *de = dictFind(channelModuleSubscribers, channel);
+    subscribers = dictGetVal(de);
+    listRewind(subscribers, &li);
+    listNode *ln;
+    while((ln = listNext(&li))) {
+        RedisModuleChannelSubscriber *sub = ln->value;
+        if (sub->module->name == module) {
+            listDelNode(subscribers, ln);
+            retval = 1;
+            zfree(sub);
+            if (listLength(subscribers) == 0) {
+                /* Free the list and associated hash entry. */
+                dictDelete(channelModuleSubscribers, channel);
+                decrRefCount(channel);
+            }
+        }
+    }
+    return retval;
+}
+
+/* Subscribe the module to a channel with a callback.  */
+int RM_SubscribeToChannel(RedisModuleCtx *ctx, RedisModuleString *channel, RedisModuleSubscribeFunc callback) {
```

Comment:
What's the behavior is you subscribe to a channel that you have previously subscribed too? Seems like you double add stuff today, which I wouldn't expect.